nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/metadata/ArtifactsXmlAbsoluteUrlRemover.java

```diff
@@ -147,52 +160,149 @@ private TempBlob transformXmlMetadata(final TempBlob artifact,
+          }
+          writer.add(xmlEvent);
+        }
+        buffer.clear();
       }
       else {
-        buffer.add(event);
+        writer.add(event);
       }
+    }
+  }
 
-      for (XMLEvent xmlEvent : buffer) {
-        writer.add(xmlEvent);
-      }
-      buffer.clear();
+  private Optional<String> getLocationAttributeFromXmlEvent(final StartElement locationElement) {
+    Attribute locationAttribute = locationElement.getAttributeByName(LOCATION_ATTR_NAME);
+    if (locationAttribute != null) {
+      return Optional.ofNullable(locationAttribute.getValue());
     }
+    return Optional.empty();
   }
 
-  private XMLEvent changeLocationAttribute(
-      final StartElement locationElement,
+  private List<String> changeLocationAttribute(
+      final String locationAttributeValue,
       final URI remoteUrl,
-      final String nexusRepositoryName)
+      final Repository nexusRepository,
+      final String file) throws IOException, XMLStreamException
+  {
+    URI uri = URI.create(locationAttributeValue);
+    return convertCompositeUrlToSimples(uri.isAbsolute() ? uri.toString() : remoteUrl.resolve(uri).toString(), nexusRepository, file).stream()
+        .map(P2PathUtils::escapeUriToPath)
+    .collect(Collectors.toList());
+  }
+
+  private List<XMLEvent> createLocationXmlEvent(String locationValue) {
+    QName childNameAttr = new QName("child");
+    return Lists.newArrayList(XML_EVENT_FACTORY.createStartElement(childNameAttr, Collections.singletonList(
+        XML_EVENT_FACTORY
+            .createAttribute(LOCATION_ATTR_NAME, locationValue))
+            .iterator(),
+        Collections.emptyIterator()), XML_EVENT_FACTORY.createEndElement(childNameAttr, Collections.singletonList(
+        XML_EVENT_FACTORY
+            .createAttribute(LOCATION_ATTR_NAME, locationValue))
+            .iterator()));
+  }
+
+  private List<String> convertCompositeUrlToSimples(final String urlString, final Repository repository, final String file)
+      throws IOException, XMLStreamException
   {
-    QName locationAttrName = new QName("location");
-    Attribute locationAttribute = locationElement.getAttributeByName(locationAttrName);
+    AtomicReference<Path> compositeContentTempFile = new AtomicReference<>(null);
+    try {
+      String baseUrl = urlString.endsWith(P2PathUtils.DIVIDER) ? urlString : urlString + P2PathUtils.DIVIDER;
+      getFilePath(baseUrl, file, ".xml").ifPresent(compositeContentTempFile::set);
+      getFilePath(baseUrl, file, ".jar").ifPresent(compositeContentTempFile::set);
 
-    if (locationAttribute == null) {
-      return locationElement;
+      if (compositeContentTempFile.get() == null) {
+        return Collections.singletonList(urlString);
+      }
+      else {
+        List<String> simpleRepositories = new ArrayList<>();
+        TempBlob tempBlob = convertFileToTempBlob(compositeContentTempFile.get(), repository);
```

Comment:
refactor the code to non create unnecessary temp blob

Reply:
ok, I created temp file for it. Because I need save downloaded file (compositeContent.jar) and then get xml file from it.